src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs

```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
+                //         Test2<T>(x);
+                Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "Test2<T>").WithArguments("C.Test2<T>(T)", "T", "T").WithLocation(11, 9)
+                );
+        }
+
+        [Fact]
+        public void ConstraintsCheck_04()
+        {
+            var src = @"
+ref struct S1
+{
+}
+
+class C
+{
+    static void Main()
+    {
+        Test2((byte)2);
+        Test3((int)3);
+        Test3(new S1());
+        Test4((long)4);
+    }
+
+    static void Test1<T>(T x) where T : allows ref struct
+    {
+        System.Console.WriteLine(""Called {0}"", typeof(T));
+    }
+
+    static void Test2<T>(T x)
+    {
+        Test1(x);
+        Test1<T>(x);
+    }
+
+    static void Test3<T>(T x) where T : allows ref struct
+    {
+        Test1(x);
+        Test1<T>(x);
+    }
+
+    static void Test4<T>(T x)
+    {
+        Test2(x);
+        Test2<T>(x);
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics, options: TestOptions.ReleaseExe);
+
+            CompileAndVerify(comp, expectedOutput: !ExecutionConditionUtil.IsMonoOrCoreClr ? null : @"
+Called System.Byte
+Called System.Byte
+Called System.Int32
+Called System.Int32
+Called S1
+Called S1
+Called System.Int64
+Called System.Int64
+Called System.Int64
+Called System.Int64", verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ConstraintsCheck_05()
+        {
+            var src = @"
+ref struct S1
+{
+}
+
+class C<T, S>
+    where T : allows ref struct
+    where S : T
+{
+    static void Main()
+    {
+        _ = typeof(C<S1, S1>);
+        _ = typeof(C<int, int>);
+        _ = typeof(C<object, object>);
+        _ = typeof(C<object, string>);
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            comp.VerifyDiagnostics(
+                // (12,26): error CS9504: The type 'S1' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter 'S' in the generic type or method 'C<T, S>'
+                //         _ = typeof(C<S1, S1>);
+                Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "S1").WithArguments("C<T, S>", "S", "S1").WithLocation(12, 26)
+                );
+        }
+
+        [Fact]
+        public void IllegalBoxing_01()
+        {
+            var src = @"
+public class Helper
+{
+    public static object Test<T>(T x) where T : allows ref struct
+    {
+        return x;
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            // PROTOTYPE(RefStructInterfaces): The boxing should be disallowed.
+            comp.VerifyDiagnostics(
+                );
+        }
```

Comment:
>} [](http://example.com/codeflow?start=8&length=1)

Consider testing explicit cast as well: `return (object)x;`.

Reply:
> Are we testing boxing for this?

Why would this make a difference? Conversion from ref struct to an interface is covered in ```ImplementAnInterface_02_IllegalBoxing```. I do not find anything interesting in the suggested scenario. If I am missing something, please let me know.